File: game_engine/models/player.py
from typing import TYPE_CHECKING, List
from game_engine.models.card import Card

if TYPE_CHECKING:
    from game_engine.logic.game import Game

class Player:
    def __init__(self, id, name, index, game: 'Game', avatar=None):
        self.id = id
        self.name = name
        self.index = index
        self.game = game  # Reference to game for dealer check
        self.hand: List[Card] = []
        self.captured_cards = []
        self.score = 0
        self.team = 'us' if index % 2 == 0 else 'them'
        self.action_text = ''
        self.last_reasoning = ''
        self.is_bot = False
        self.avatar = avatar if avatar else f"https://picsum.photos/id/{64 + (index % 4)}/100/100"
        
        # Director Configs (Bot Instructions)
        self.strategy = 'heuristic' # default
        self.profile = None # Explicit personality override
        self.difficulty = None # Difficulty level (EASY/MEDIUM/HARD/KHALID)

    @property
    def position(self):
        positions = ['Bottom', 'Right', 'Top', 'Left']
        return positions[self.index]

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "avatar": self.avatar,
            "index": self.index,
            "hand": [c.to_dict() for c in self.hand],
            "score": self.score,
            "team": self.team,
            "position": self.position,
            "isDealer": (self.index == self.game.dealer_index),
            "actionText": self.action_text,
            "lastReasoning": self.last_reasoning,
            "isBot": self.is_bot, 
            "isActive": (self.index == self.game.current_turn),
            # Expose Configs
            "strategy": getattr(self, 'strategy', 'heuristic'),
            "profile": getattr(self, 'profile', None),
            "difficulty": getattr(self, 'difficulty', None)
        }

File: mobile/lib/models/player.dart
/// Player — A game participant (human or bot).
///
/// Port of frontend/src/types.ts Player interface.
library;
import 'card_model.dart';
import 'enums.dart';

class Player {
  final PlayerPosition position;
  final String name;
  final String avatar;
  final List<CardModel> hand;
  final int score;
  final bool isDealer;
  final bool isActive;
  final String? actionText;
  final String? lastReasoning;
  final int index;
  final bool isBot;

  const Player({
    required this.position,
    required this.name,
    this.avatar = '',
    this.hand = const [],
    this.score = 0,
    this.isDealer = false,
    this.isActive = false,
    this.actionText,
    this.lastReasoning,
    required this.index,
    this.isBot = false,
  });

  Player copyWith({
    PlayerPosition? position,
    String? name,
    String? avatar,
    List<CardModel>? hand,
    int? score,
    bool? isDealer,
    bool? isActive,
    String? actionText,
    String? lastReasoning,
    int? index,
    bool? isBot,
  }) {
    return Player(
      position: position ?? this.position,
      name: name ?? this.name,
      avatar: avatar ?? this.avatar,
      hand: hand ?? this.hand,
      score: score ?? this.score,
      isDealer: isDealer ?? this.isDealer,
      isActive: isActive ?? this.isActive,
      actionText: actionText ?? this.actionText,
      lastReasoning: lastReasoning ?? this.lastReasoning,
      index: index ?? this.index,
      isBot: isBot ?? this.isBot,
    );
  }

  factory Player.fromJson(Map<String, dynamic> json) {
    return Player(
      position: PlayerPosition.fromValue(json['position'] as String? ?? 'Bottom'),
      name: json['name'] as String? ?? '',
      avatar: json['avatar'] as String? ?? '',
      hand: (json['hand'] as List<dynamic>?)
              ?.map((c) => CardModel.fromJson(c as Map<String, dynamic>))
              .toList() ??
          [],
      score: json['score'] as int? ?? 0,
      isDealer: json['isDealer'] as bool? ?? false,
      isActive: json['isActive'] as bool? ?? false,
      actionText: json['actionText'] as String?,
      lastReasoning: json['lastReasoning'] as String?,
      index: json['index'] as int? ?? 0,
      isBot: json['is_bot'] as bool? ?? json['isBot'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
        'position': position.value,
        'name': name,
        'avatar': avatar,
        'hand': hand.map((c) => c.toJson()).toList(),
        'score': score,
        'isDealer': isDealer,
        'isActive': isActive,
        if (actionText != null) 'actionText': actionText,
        if (lastReasoning != null) 'lastReasoning': lastReasoning,
        'index': index,
        'isBot': isBot,
      };
}
